extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/registry/json/JsonExporter.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.quarkus.micrometer.runtime.registry.json;
+
+import java.io.StringWriter;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import javax.json.JsonObject;
+import javax.json.JsonObjectBuilder;
+import javax.json.JsonValue;
+import javax.json.JsonWriter;
+import javax.json.spi.JsonProvider;
+import javax.json.stream.JsonGenerator;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.FunctionCounter;
+import io.micrometer.core.instrument.FunctionTimer;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.core.instrument.distribution.HistogramSnapshot;
+import io.micrometer.core.instrument.distribution.ValueAtPercentile;
+
+public class JsonExporter {
+
+    private static final Map<String, ?> JSON_CONFIG = Collections.singletonMap(JsonGenerator.PRETTY_PRINTING, true);
+    private static final JsonProvider JSON_PROVIDER = JsonProvider.provider();
+
+    public JsonExporter() {
+    }
+
+    public StringBuilder exportEverything(JsonMeterRegistry meterRegistry) {
+        JsonObjectBuilder root = JSON_PROVIDER.createObjectBuilder();
+        exportCounters(retrieveFromCollection(Counter.class, meterRegistry)).forEach(root::add);
+        exportGauges(retrieveFromCollection(Gauge.class, meterRegistry)).forEach(root::add);
+        exportFunctionCounters(retrieveFromCollection(FunctionCounter.class, meterRegistry)).forEach(root::add);
+        exportTimers(retrieveFromCollection(Timer.class, meterRegistry)).forEach(root::add);
+        exportFunctionTimers(retrieveFromCollection(FunctionTimer.class, meterRegistry)).forEach(root::add);
+        exportDistributionSummaries(retrieveFromCollection(DistributionSummary.class, meterRegistry)).forEach(root::add);
+        return stringify(root.build());
```

Comment:
Given what this does (retrieve from collection effectively retrieves from the stream repeatedly), I think you should rejigger this to use the match function on the Meter interface as is done here: 
https://github.com/micrometer-metrics/micrometer/blob/master/implementations/micrometer-registry-appoptics/src/main/java/io/micrometer/appoptics/AppOpticsMeterRegistry.java#L98
(most other registries do this as well).

Gives you a one time loop through everything that you can use to initialize those collections with what you need, and then you can emit the lot at the end.

Reply:
Given that the `Meter.match` method takes functions that act on only one meter at a time (and I need to group the meters by name and type first to be able to export), I don't see any other use for it than for potentially performing the grouping itself, but it would feel quite awkward and we can do that more effectively using pure Java streams.. So I guess I'd need more hints how you imagine doing this?!

Reply:
I could imagine this
```
        List<Gauge> gauges = new ArrayList<>();
        List<Counter> counters = new ArrayList<>();
        List<TimeGauge> timeGauges = new ArrayList<>();
        List<FunctionCounter> functionCounters = new ArrayList<>();
        List<Timer> timers = new ArrayList<>();
        List<LongTaskTimer> longTaskTimers = new ArrayList<>();
        List<FunctionTimer> functionTimers = new ArrayList<>();
        List<DistributionSummary> distributionSummaries = new ArrayList<>();
        List<Meter> meters = new ArrayList<>();
        meterRegistry.getMeters().forEach(meter -> meter.match(gauges::add,
                counters::add,
                timers::add,
                distributionSummaries::add,
                longTaskTimers::add,
                timeGauges::add,
                functionCounters::add,
                functionTimers::add,
                meters::add));
        exportCounters(counters).forEach(root::add);
        exportGauges(gauges).forEach(root::add);
        exportTimeGauges(timeGauges).forEach(root::add);
        exportFunctionCounters(functionCounters).forEach(root::add);
        exportTimers(timers).forEach(root::add);
        exportLongTaskTimers(longTaskTimers).forEach(root::add);
        exportFunctionTimers(functionTimers).forEach(root::add);
        exportDistributionSummaries(distributionSummaries).forEach(root::add);
```
but perhaps you had something better in mind

Reply:
Well. The way it exists now, you're iterating the list of registered meters several times: `registry.getMeters().stream()` in retrieve from collection. The above does that iteration/collection in one fell swoop.

Reply:
Ok, I committed this change.